drivers: modem_cellular: changes due to net_link_addr struct changes

jhirsi opened the pull request:
Recent changes to net_link_addr structure to not use pointers
have made it necessary to update the modem_cellular driver
to ensure compatibility with the new structure definition.
Otherwise, an assertion failure occurs in subsys/net/l2/ppp/ipv6cp.c:40.
This commit updates that link address is set to NET_LINK_ADDR_MAX_LENGTH
least significant bytes of IMEI instead of IMEI total length.

jukkar reviewed (changes requested):
Adding -1 for now to avoid accidental merging.

jukkar commented on include/zephyr/net/net_linkaddr.h:
```diff
@@ -30,7 +30,9 @@ extern "C" {
 /** Maximum length of the link address */
-#if defined(CONFIG_NET_L2_PHY_IEEE802154) || defined(CONFIG_NET_L2_PPP)
+#if defined(CONFIG_MODEM_CELLULAR)
+#define NET_LINK_ADDR_MAX_LENGTH 16
```
Is it possible to add a test for this so that we can verify that at least the tests work?

jukkar replied:
> It is real issue with ipv6 and when using NCS SLM application from ext-mcu and in this case from cellular modem.

I am sure things work ok if only using cellular modem. I am mostly concerned what happens with multiple interface scenario like having wifi enabled at the same time.

jukkar reviewed (commented):
I did some manual testing by setting link address to 16 bytes. Everything seems to work ok with Ethernet.

rlubos commented:
Just my 2c, `MODEM_CELLULAR` seems to be dependent on `NET_L2_PPP`, so IMHO it should comply with whatever LL address size the PPP L2 specifies (which currently is 8 bytes for whatever reason).

jhirsi commented on drivers/modem/modem_cellular.c:
```diff
@@ -881,8 +882,12 @@ static void modem_cellular_run_init_script_event_handler(struct modem_cellular_d
 	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
-		net_if_set_link_addr(modem_ppp_get_iface(data->ppp), data->imei,
-				     ARRAY_SIZE(data->imei), NET_LINK_UNKNOWN);
+		err = net_if_set_link_addr(modem_ppp_get_iface(data->ppp), data->imei,
+				     ARRAY_SIZE(data->imei),
+				     NET_LINK_UNKNOWN);
+		if (err) {
+			LOG_WRN("Failed to set link address on PPP interface (%d)", err);
+		}
 
```
Good idea. I've done these changes now and pushing changes shortly. I'll change the topic of the PR accordingly.

MarkusLassila commented on drivers/modem/modem_cellular.c:
```diff
@@ -881,8 +881,18 @@ static void modem_cellular_run_init_script_event_handler(struct modem_cellular_d
+		/* Get link_addr_len least significant bytes from IMEI as a link address */
+		int link_addr_len = MIN(NET_LINK_ADDR_MAX_LENGTH, MODEM_CELLULAR_DATA_IMEI_LEN);
+		uint8_t *link_addr_ptr = data->imei +
+			(MODEM_CELLULAR_DATA_IMEI_LEN - link_addr_len - 1); /* Exclude str end */
```
If MODEM_CELLULAR_DATA_IMEI_LEN == NET_LINK_ADDR_MAX_LENGTH, link_addr_ptr would point one byte before data->imei.

jhirsi committed:
drivers: modem_cellular: changes due to net_link_addr struct changes

Recent changes to net_link_addr structure to not use pointers
have made it necessary to update the modem_cellular driver
to ensure compatibility with the new structure definition.
Otherwise, an assertion failure occurs in subsys/net/l2/ppp/ipv6cp.c:40.
This commit updates that link address is set to NET_LINK_ADDR_MAX_LENGTH
least significant bytes of IMEI instead of IMEI total length.

Signed-off-by: Jani Hirsimäki <jani.hirsimaki@nordicsemi.no>
